lib/charms/opensearch/v0/opensearch_snapshots.py

```diff
@@ -0,0 +1,1715 @@
+        charm: "OpenSearchBaseCharm",
+    ):
+        super().__init__(charm, key="backups")
+        self.charm = charm
+
+        # requirers
+        self.s3_requirer = S3Requirer(charm, S3_RELATION)
+        self.azure_requirer = AzureStorageRequires(charm, AZURE_RELATION)
+
+        # simple deployments or main orchestrator
+        self.framework.observe(
+            self.s3_requirer.on.credentials_changed, self._on_s3_credentials_changed
+        )
+        self.framework.observe(self.s3_requirer.on.credentials_gone, self._on_s3_credentials_gone)
+        self.framework.observe(
+            self.azure_requirer.on.storage_connection_info_changed,
+            self._on_azure_credentials_changed,
+        )
+        self.framework.observe(
+            self.azure_requirer.on.storage_connection_info_gone, self._on_azure_credentials_gone
+        )
+
+        # large deployments with non-main orchestrator
+        self.framework.observe(
+            charm.on[PeerClusterRelationName].relation_changed,
+            self._on_peer_clusters_relation_changed_for_snapshots,
+        )
+        self.framework.observe(
+            charm.on[PeerClusterRelationName].relation_departed,
+            self._on_peer_clusters_relation_departed_for_snapshots,
+        )
+
+        # actions
+        self.framework.observe(self.charm.on.create_backup_action, self._on_create_backup_action)
+        self.framework.observe(self.charm.on.list_backups_action, self._on_list_backups_action)
+        self.framework.observe(self.charm.on.restore_action, self._on_restore_action)
+
+    def _on_s3_credentials_changed(self, event: CredentialsChangedEvent) -> None:  # noqa: C901
+        """Handler for s3 credentials changed event."""
+        if not (deployment_desc := self.charm.opensearch_peer_cm.deployment_desc()):
+            logger.debug("Deployment description not ready; deferring %s", event)
+            event.defer()
+            return
+        # block non-main orchestrators only when they are in a multi-app topology.
+        if (
+            deployment_desc.typ != DeploymentType.MAIN_ORCHESTRATOR
+            and self._is_part_of_large_deployment()
+        ):
+            if self.charm.unit.is_leader():
+                self.charm.status.set(BlockedStatus(BackupRelShouldNotExist), app=True)
+            return
+
+        object_storage_type = self.charm.snapshots_manager.get_storage_type()
+        if not object_storage_type:
+            return
+
+        logger.info(f"S3 credentials changed for object storage type {object_storage_type}")
+        if object_storage_type == "conflict":
+            if self.charm.unit.is_leader():
+                self.charm.status.set(BlockedStatus(BackupRelConflict), app=True)
+            event.defer()
+            return
+
+        if self.charm.unit.is_leader():
+            self.charm.status.clear(BackupRelConflict, app=True)
+
+        cfg = self.charm.snapshots_manager.get_storage_config(ObjectStorageType.S3)
```

Comment:
nit, for readability 
```suggestion
        object_storage_config = self.charm.snapshots_manager.get_storage_config(ObjectStorageType.S3)
```

Reply:
Done